File: python/zyvor_janus/server/openai_shim.py
"""OpenAI-compatible virtual endpoint backed by Zyvor Janus inference model."""

from __future__ import annotations

import asyncio
import json
import os
import time
import uuid
from collections import defaultdict
from typing import Any, AsyncIterator

from fastapi import APIRouter, Depends, Header, HTTPException, Request
from fastapi.responses import StreamingResponse
from pydantic import BaseModel, Field

from zyvor_janus.adapters.profiles import ProfileRegistry

router = APIRouter(prefix="/v1", tags=["openai-shim"])

DEFAULT_API_KEY = os.environ.get("ZYVOR_JANUS_API_KEY", "dev-zyvor-janus-key")
RATE_LIMIT_PER_MIN = int(os.environ.get("ZYVOR_JANUS_SHIM_RATE_LIMIT", "120"))
PROFILES_DIR = os.environ.get("ZYVOR_JANUS_PROFILES_DIR", "configs/profiles")

_rate_buckets: dict[str, list[float]] = defaultdict(list)
_profile_registry = ProfileRegistry(__import__("pathlib").Path(PROFILES_DIR))


class ChatMessage(BaseModel):
    role: str
    content: str


class ChatCompletionRequest(BaseModel):
    model: str
    messages: list[ChatMessage] = Field(default_factory=list)
    stream: bool = False
    max_tokens: int = 128


def _estimate_tokens(messages: list[ChatMessage]) -> tuple[int, int]:
    input_tokens = sum(max(1, len(m.content.split())) for m in messages)
    return input_tokens, max(16, min(512, input_tokens // 2))


def _require_api_key(authorization: str | None = Header(default=None)) -> str:
    if not authorization or not authorization.startswith("Bearer "):
        raise HTTPException(status_code=401, detail="missing bearer token")
    token = authorization.removeprefix("Bearer ").strip()
    if token != DEFAULT_API_KEY:
        raise HTTPException(status_code=401, detail="invalid api key")
    return token


def _check_rate_limit(client_key: str) -> None:
    now = time.time()
    window = _rate_buckets[client_key]
    _rate_buckets[client_key] = [t for t in window if now - t < 60.0]
    if len(_rate_buckets[client_key]) >= RATE_LIMIT_PER_MIN:
        raise HTTPException(status_code=429, detail="rate limit exceeded")
    _rate_buckets[client_key].append(now)


def _estimate_ttft_ms(model: str, input_tokens: int, output_tokens: int) -> float:
    try:
        prefill_ms, decode_tps = _profile_registry.lookup_v2(model, "H100")
    except Exception:
        prefill_ms, decode_tps = 0.08, 120.0
    ttft = input_tokens * prefill_ms
    decode_secs = output_tokens / max(decode_tps, 1.0)
    return ttft + decode_secs * 1000.0 * 0.05


async def _stream_tokens(content: str, ttft_ms: float) -> AsyncIterator[str]:
    await asyncio.sleep(ttft_ms / 1000.0)
    words = content.split()
    if not words:
        words = ["OK"]
    for word in words:
        chunk = {
            "id": f"chatcmpl-{uuid.uuid4().hex[:8]}",
            "object": "chat.completion.chunk",
            "choices": [{"index": 0, "delta": {"content": word + " "}, "finish_reason": None}],
        }
        yield f"data: {json.dumps(chunk)}\n\n"
        await asyncio.sleep(0.02)
    done = {
        "id": f"chatcmpl-{uuid.uuid4().hex[:8]}",
        "object": "chat.completion.chunk",
        "choices": [{"index": 0, "delta": {}, "finish_reason": "stop"}],
    }
    yield f"data: {json.dumps(done)}\n\n"
    yield "data: [DONE]\n\n"


@router.post("/chat/completions")
async def chat_completions(
    body: ChatCompletionRequest,
    request: Request,
    _token: str = Depends(_require_api_key),
) -> Any:
    client_key = request.client.host if request.client else "local"
    _check_rate_limit(client_key)
    input_tokens, default_output = _estimate_tokens(body.messages)
    output_tokens = body.max_tokens or default_output
    ttft_ms = _estimate_ttft_ms(body.model, input_tokens, output_tokens)
    reply = " ".join(m.content for m in body.messages if m.role == "assistant") or "Zyvor Janus virtual completion."
    if body.stream:
        return StreamingResponse(_stream_tokens(reply, ttft_ms), media_type="text/event-stream")
    await asyncio.sleep(ttft_ms / 1000.0)
    return {
        "id": f"chatcmpl-{uuid.uuid4().hex[:8]}",
        "object": "chat.completion",
        "model": body.model,
        "choices": [
            {
                "index": 0,
                "message": {"role": "assistant", "content": reply},
                "finish_reason": "stop",
            }
        ],
        "usage": {
            "prompt_tokens": input_tokens,
            "completion_tokens": output_tokens,
            "total_tokens": input_tokens + output_tokens,
        },
    }

File: crates/zyvor-janus-api/src/routes/openai_shim.rs
// Copyright 2026 ZyvorAI Labs Private Limited
// SPDX-License-Identifier: Apache-2.0

use std::net::SocketAddr;
use std::time::{Duration, Instant};

use async_stream::stream;
use axum::extract::{ConnectInfo, Request, State};
use axum::http::header::AUTHORIZATION;
use axum::middleware::Next;
use axum::response::sse::{Event, Sse};
use axum::response::{IntoResponse, Response};
use axum::Json;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::error::ApiError;
use crate::state::AppState;

#[derive(Deserialize)]
pub struct ChatMessage {
    role: String,
    content: String,
}

#[derive(Deserialize)]
pub struct ChatCompletionRequest {
    model: String,
    #[serde(default)]
    messages: Vec<ChatMessage>,
    #[serde(default)]
    stream: bool,
    #[serde(default = "default_max_tokens")]
    max_tokens: i64,
}

fn default_max_tokens() -> i64 {
    128
}

/// Port of Python's `_require_api_key` -- deliberately its own check, not
/// the blanket `/api/*` bearer middleware (auth.rs), so this route's error
/// text and credential handling stay independent even though today both
/// happen to check the same `ZYVOR_JANUS_API_KEY`.
///
/// A middleware rather than an in-handler check: FastAPI resolves
/// `Depends(_require_api_key)` ahead of Pydantic body validation, so an
/// unauthenticated request with a malformed body still gets 401, not a
/// validation error. Axum extractors (including `Json<T>`) run in the
/// handler's declared order and short-circuit on the first failure, so an
/// in-handler check after a `Json<T>` param would see 422 win that race
/// instead -- a middleware runs before any of the handler's extractors.
pub async fn require_shim_api_key(
    State(state): State<AppState>,
    request: Request,
    next: Next,
) -> Result<Response, ApiError> {
    let auth_header = request
        .headers()
        .get(AUTHORIZATION)
        .and_then(|v| v.to_str().ok());
    let Some(auth_header) = auth_header else {
        return Err(ApiError::unauthorized("missing bearer token"));
    };
    let Some(token) = auth_header.strip_prefix("Bearer ") else {
        return Err(ApiError::unauthorized("missing bearer token"));
    };
    if !constant_time_eq(token.trim().as_bytes(), state.inner.api_key.as_bytes()) {
        return Err(ApiError::unauthorized("invalid api key"));
    }
    Ok(next.run(request).await)
}

fn constant_time_eq(a: &[u8], b: &[u8]) -> bool {
    if a.len() != b.len() {
        return false;
    }
    let mut diff = 0u8;
    for (x, y) in a.iter().zip(b.iter()) {
        diff |= x ^ y;
    }
    diff == 0
}

/// Port of Python's `_check_rate_limit`: a 60-second sliding window per
/// client key, capped at `ZYVOR_JANUS_SHIM_RATE_LIMIT` requests/min.
fn check_rate_limit(state: &AppState, client_key: &str) -> Result<(), ApiError> {
    let now = Instant::now();
    let mut buckets = state.inner.rate_buckets.lock().expect("rate bucket lock");
    let window = buckets.entry(client_key.to_string()).or_default();
    window.retain(|t| now.duration_since(*t) < Duration::from_secs(60));
    if window.len() >= state.inner.shim_rate_limit_per_min as usize {
        return Err(ApiError::too_many_requests("rate limit exceeded"));
    }
    window.push(now);
    Ok(())
}

/// Port of Python's `_estimate_tokens`: input = sum of whitespace-split word
/// counts (min 1 per message) across every message regardless of role;
/// default output = clamp(input / 2, 16, 512).
fn estimate_tokens(messages: &[ChatMessage]) -> (i64, i64) {
    let input_tokens: i64 = messages
        .iter()
        .map(|m| (m.content.split_whitespace().count() as i64).max(1))
        .sum();
    let default_output = (input_tokens / 2).clamp(16, 512);
    (input_tokens, default_output)
}

fn estimate_ttft_ms(state: &AppState, model: &str, input_tokens: i64, output_tokens: i64) -> f64 {
    let (prefill_ms, decode_tps) = state.inner.model_profiles.lookup_v2(model, "H100");
    let ttft = input_tokens as f64 * prefill_ms;
    let decode_secs = output_tokens as f64 / decode_tps.max(1.0);
    ttft + decode_secs * 1000.0 * 0.05
}

fn chat_id() -> String {
    format!("chatcmpl-{}", &Uuid::new_v4().simple().to_string()[..8])
}

#[derive(Serialize)]
struct ChatChunkDelta {
    #[serde(skip_serializing_if = "Option::is_none")]
    content: Option<String>,
}

#[derive(Serialize)]
struct ChatChunkChoice {
    index: u32,
    delta: ChatChunkDelta,
    finish_reason: Option<&'static str>,
}

#[derive(Serialize)]
struct ChatChunk {
    id: String,
    object: &'static str,
    choices: Vec<ChatChunkChoice>,
}

/// Port of Python's `_stream_tokens`: sleeps `ttft_ms`, then emits one SSE
/// chunk per whitespace-split word (20ms apart), a final `finish_reason:
/// "stop"` chunk, then the `[DONE]` sentinel.
fn stream_tokens(
    content: String,
    ttft_ms: f64,
) -> impl futures_core::Stream<Item = Result<Event, std::convert::Infallible>> {
    stream! {
        tokio::time::sleep(Duration::from_secs_f64((ttft_ms / 1000.0).max(0.0))).await;
        let mut words: Vec<&str> = content.split_whitespace().collect();
        if words.is_empty() {
            words.push("OK");
        }
        for word in words {
            let chunk = ChatChunk {
                id: chat_id(),
                object: "chat.completion.chunk",
                choices: vec![ChatChunkChoice {
                    index: 0,
                    delta: ChatChunkDelta { content: Some(format!("{word} ")) },
                    finish_reason: None,
                }],
            };
            yield Ok(Event::default().data(serde_json::to_string(&chunk).unwrap_or_default()));
            tokio::time::sleep(Duration::from_millis(20)).await;
        }
        let done_chunk = ChatChunk {
            id: chat_id(),
            object: "chat.completion.chunk",
            choices: vec![ChatChunkChoice {
                index: 0,
                delta: ChatChunkDelta { content: None },
                finish_reason: Some("stop"),
            }],
        };
        yield Ok(Event::default().data(serde_json::to_string(&done_chunk).unwrap_or_default()));
        yield Ok(Event::default().data("[DONE]"));
    }
}

#[derive(Serialize)]
struct ChatMessageOut {
    role: &'static str,
    content: String,
}

#[derive(Serialize)]
struct ChatChoice {
    index: u32,
    message: ChatMessageOut,
    finish_reason: &'static str,
}

#[derive(Serialize)]
struct Usage {
    prompt_tokens: i64,
    completion_tokens: i64,
    total_tokens: i64,
}

#[derive(Serialize)]
struct ChatCompletionResponse {
    id: String,
    object: &'static str,
    model: String,
    choices: Vec<ChatChoice>,
    usage: Usage,
}

/// `POST /v1/chat/completions` -- OpenAI-compatible virtual endpoint backed
/// by no real model; timing is simulated via `estimate_ttft_ms`, the "reply"
/// echoes any `assistant`-role messages in the request (or a fixed fallback
/// string), matching Python's `openai_shim.py` exactly, hallucination and
/// all -- this endpoint never calls a real LLM in either implementation.
pub async fn chat_completions(
    State(state): State<AppState>,
    ConnectInfo(addr): ConnectInfo<SocketAddr>,
    Json(body): Json<ChatCompletionRequest>,
) -> Result<Response, ApiError> {
    check_rate_limit(&state, &addr.ip().to_string())?;

    let (input_tokens, default_output) = estimate_tokens(&body.messages);
    let output_tokens = if body.max_tokens != 0 {
        body.max_tokens
    } else {
        default_output
    };
    let ttft_ms = estimate_ttft_ms(&state, &body.model, input_tokens, output_tokens);
    let reply = {
        let joined = body
            .messages
            .iter()
            .filter(|m| m.role == "assistant")
            .map(|m| m.content.as_str())
            .collect::<Vec<_>>()
            .join(" ");
        if joined.is_empty() {
            "Zyvor Janus virtual completion.".to_string()
        } else {
            joined
        }
    };

    if body.stream {
        return Ok(Sse::new(stream_tokens(reply, ttft_ms)).into_response());
    }

    tokio::time::sleep(Duration::from_secs_f64((ttft_ms / 1000.0).max(0.0))).await;
    Ok(Json(ChatCompletionResponse {
        id: chat_id(),
        object: "chat.completion",
        model: body.model,
        choices: vec![ChatChoice {
            index: 0,
            message: ChatMessageOut {
                role: "assistant",
                content: reply,
            },
            finish_reason: "stop",
        }],
        usage: Usage {
            prompt_tokens: input_tokens,
            completion_tokens: output_tokens,
            total_tokens: input_tokens + output_tokens,
        },
    })
    .into_response())
}
